engine/jobmaster/dm/dm_jobmaster.go

```diff
@@ -118,26 +112,41 @@ func (jm *JobMaster) InitImpl(ctx context.Context) error {
 	if err := jm.initComponents(ctx); err != nil {
 		return err
 	}
-	if err := jm.preCheck(ctx, jm.initJobCfg); err != nil {
-		return err
+	jm.mu.RLock()
+	cfg := jm.cfg
+	jm.mu.RUnlock()
+	if err := jm.preCheck(ctx, cfg); err != nil {
+		return jm.Exit(ctx, frameModel.WorkerStatus{Code: frameModel.WorkerStatusError}, err)
 	}
-	if err := jm.checkpointAgent.Create(ctx); err != nil {
+	if err := jm.checkpointAgent.Create(ctx, cfg); err != nil {
 		return err
 	}
-	return jm.taskManager.OperateTask(ctx, dmpkg.Create, jm.initJobCfg, nil)
+	return jm.taskManager.OperateTask(ctx, dmpkg.Create, cfg, nil)
 }
 
 // Tick implements JobMasterImpl.Tick
 func (jm *JobMaster) Tick(ctx context.Context) error {
 	jm.workerManager.Tick(ctx)
 	jm.taskManager.Tick(ctx)
-	return jm.messageAgent.Tick(ctx)
+	if err := jm.messageAgent.Tick(ctx); err != nil {
+		return err
+	}
+	if jm.isFinished(ctx) {
+		return jm.cancel(ctx, frameModel.WorkerStatusFinished)
+	}
+	return nil
 }
 
 // OnMasterRecovered implements JobMasterImpl.OnMasterRecovered
+// When it is called, the jobCfg may not be in the metadata, and we should not report an error
```

Comment:
seems we don't check if jm.cfg is nil in InitImpl. And the comment doesn't indicate the relation of OnMasterRecovered and InitImpl. Please provide a document for calling sequence @liuzix